Add checks to make the cinder services more reliable

rajathere opened the pull request:
This PR has 5 commits:
1. use shell_header instead of custom 'set -eux' command
2. Move the cinder endpoint check to beginning (before fetching old services)
3. Wait for cinder-scheduler to report 'up' in cinder services DB
4. Run the online data migration from cinder-api (to be consistent)
5. update the service 'up' check to use raw strings instead of grepping from OSC output

Jira: https://issues.redhat.com/browse/OSPCIX-564

karelyatin commented on tests/roles/cinder_adoption/tasks/main.yaml:
```diff
@@ -14,13 +14,27 @@
-    set -eux
     oc exec -t openstackclient -- openstack volume service list --service cinder-backup -c Host -f value
   register: backup_down_services
+  retries: 10
```
let's not add retries here

rajathere replied:
Done

karelyatin commented on tests/roles/cinder_adoption/tasks/main.yaml:
```diff
@@ -29,13 +43,14 @@
-    set -eux
     oc exec -t openstackclient -- openstack volume service list --service cinder-scheduler -c Host -f value
   register: scheduler_down_services
+  retries: 10
```
same

rajathere replied:
Done

karelyatin commented:
/lgtm

rajathere commented:
Verification:

1. endpoint check happens before getting services for removal

TASK [cinder_adoption : check that Cinder is reachable and its endpoints are defined] ***
...
42f3-80ac-dab53f34f06b | tripleo | True      |", "+--------------------------------------+---------+-----------+"]}

TASK [cinder_adoption : Get cinder-backup down services] ***********************

2. pipefail included in the tasks

TASK [cinder_adoption : Get cinder-backup down services] ***********************
changed: [localhost] => {"changed": true, "cmd": "set -euxo pipefail\n\n\noc exec -t ...

3. Waiting for cinder-scheduler service to be up

TASK [cinder_adoption : wait for Cinder scheduler to be up and ready] **********
changed: [localhost] => {"attempts": 1, "changed": true, "cmd": "set -euxo pipefail\n\n\noc exec -t openstackclient -- openstack volume service list --service cinder-scheduler -c State -f value | grep 'up'\n", "delta": "0:00:02.991715", "end": "2024-11-21 12:25:49.292114", "msg": "", "rc": 0, "start": "2024-11-21 12:25:46.300399", "stderr": "+ oc exec -t openstackclient -- openstack volume service list --service cinder-scheduler -c State -f value\n+ grep up", "stderr_lines": ["+ oc exec -t openstackclient -- openstack volume service list --service cinder-scheduler -c State -f value", "+ grep up"], "stdout": "up", "stdout_lines": ["up"]}

4. online data migrations are run in cinder-api container

oc exec -t cinder-api-0 -c cinder-api -- cinder-manage db online_data_migrations

5. output raw strings vs grep from OSC output

"+ oc exec -t openstackclient -- openstack volume service list --service cinder-backup -c State -f value\n+ grep up", "stderr_lines": ["+ oc exec -t openstackclient -- openstack volume service list --service cinder-backup -c State -f value", "+ grep up"], "stdout": "up", "stdout_lines": ["up"]}

karelyatin commented:
Also validated in
periodic-adoption-multinode-to-crc-no-ceph-rollback https://softwarefactory-project.io/zuul/t/rdoproject.org/build/3a89a2c6cd4846e7b3fb688f12d9bcf6 : SUCCESS in 3h 07m 04s
periodic-adoption-standalone-to-crc-no-ceph https://softwarefactory-project.io/zuul/t/rdoproject.org/build/825488ca2f9547aa9b74cda3a3ef848d : SUCCESS in 3h 18m 34s

fmount commented on tests/roles/cinder_adoption/tasks/main.yaml:
```diff
@@ -91,25 +101,22 @@
+    oc exec -t openstackclient -- openstack volume service list --service cinder-scheduler -c State -f value | grep 'up'
+  register: cinder_scheduler_running_result
+  until: cinder_scheduler_running_result is success
+  retries: 5
```
I assume retries / delay values are sufficient according to your tests.
In case we face issues we can think about moving  `5` and `2` as defaults, so we can always tune them without requiring any code change (as they might just be overrides passed to the jobs).

rajathere replied:
The values (retries:5, delay:2) are taken from the cinder-volume and cinder-backup service state checks.
Though these work good for this specific check, other checks like waiting for the pod to be running or endpoint being reachable differs.
But I agree that configuring retries and delays for similar type of tasks would be a useful thing.

fmount reviewed (approved):
One small comment about some defaults. However they're not too much relevant with the purpose of this change and can be eventually improved later.

/lgtm

holser commented:
/cherry-pick 18.0-fr1

holser commented:
/cherry-pick 18.0-fr1
